src/mqtt_client.rs

```diff
@@ -350,12 +366,23 @@ where
             topic, data, properties
         );
 
-        let mut buffer: [u8; T] = [0; T];
-        let packet = serialize::publish_message(&mut buffer, topic, data, properties)?;
-        self.write(packet)
+        // If QoS 0 the ID will be ignored
+        let id = self.session_state.get_packet_identifier();
+
+        let mut buffer: [u8; MSG_SIZE] = [0; MSG_SIZE];
+        let packet = serialize::publish_message(&mut buffer, topic, data, qos, id, properties)?;
+
+        self.write(packet)?;
+        self.session_state.increment_packet_identifier();
+
+        if qos == QoS::AtLeastOnce {
+            self.session_state.publish_at_least_once(id, packet);
```

Comment:
For republish, I think packets need to have the DUP flag set in the PUBLISH fixed header

Reply:
Fixed as described in the comment about qos and fixed header serialization.